Add capital cost function for crystallizer model

OOAmusat opened the pull request:
## Summary/Motivation:

This PR adds two captal cost estimation functions for the crystallizer model.


## Changes proposed in this PR:
- Add mass- and volume-based capital cost estimates for crystallizer

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

OOAmusat committed:
Add documentation files for crystallization

OOAmusat committed:
Run black on crystallizer

OOAmusat committed:
Merge branch 'main' into crystallization

OOAmusat committed:
Merge branch 'main' into crystallization

OOAmusat committed:
Typo fixed.

OOAmusat committed:
Merge branch 'main' into crystallization

OOAmusat committed:
Merge branch 'crystallization' of https://github.com/OOAmusat/watertap into crystallization

OOAmusat committed:
First attempt at costing

OOAmusat committed:
Fix costing issues

OOAmusat committed:
Cost conversion to base year

OOAmusat committed:
Add volume-based costing

OOAmusat committed:
Add volume-based costing arguments

OOAmusat committed:
Add crystallizer costing tests

OOAmusat committed:
Merge branch 'watertap-org:main' into crystallization

OOAmusat committed:
black formatting

OOAmusat committed:
Merge branch 'crystallization' of https://github.com/OOAmusat/watertap into crystallization

TimBartholomew commented:
@bknueven can you look into these test failures for LSRRO? It's weird that they are failing for the number of variables and Mayo didn't touch the LSRRO files.

adam-a-a committed:
Merge branch 'main' into crystallization

adam-a-a committed:
Merge branch 'crystallization' of https://github.com/OOAmusat/watertap into crystallization

TimBartholomew commented:
@OOAmusat 
Ben's response was:
It’s these lines: https://github.com/watertap-org/watertap/pull/515/files#diff-b8409e0826e208143b2dc914e86c9f79da1e396679bd80ab49edb430ead318e1R201-R236

These new parameters appear on every costing block. Therefore we should expect an increase in the number of Vars for any flowsheet using the watertap costing package.

bknueven commented on watertap/costing/watertap_costing_package.py:
```diff
@@ -586,6 +629,67 @@ def cost_caoh2_mixer(blk):
+                    ),
+                    to_units=blk.costing_package.base_currency,
+                )
+            )
```
I've generally written different costing types as their own method (see `cost_pump` for example) for maximal flexibility. I.e. you can specify the method directly *or* use the `CrystallizerCostType` attribute.

OOAmusat replied:
Done.

bknueven commented on watertap/unit_models/tests/test_crystallizer.py:
```diff
@@ -286,6 +290,19 @@ def test_calculate_scaling(self, Crystallizer_frame):
 
     @pytest.mark.component
     def test_initialize(self, Crystallizer_frame):
+        # Add costing function, then initialize
+        m = Crystallizer_frame
+        m.fs.costing = WaterTAPCosting()
+        m.fs.unit.costing = UnitModelCostingBlock(
+            default={
+                "flowsheet_costing_block": m.fs.costing,
+                "costing_method_arguments": {
+                    "cost_type": CrystallizerCostType.mass_basis
+                },
+            },
+        )
+        m.fs.costing.cost_process()
+
```
Are these tests in the right place?

On one hand, our general philosophy has been to put tests in the same directory as the code, i.e., in the `watertap/costing/tests/` directory. On the other hand, breaking them out into the unit model tests makes a huge amount of sense as well -- we already have the code to set up the unit model, etc.

OOAmusat replied:
I could do this. However, are we sure that is how we want to handle the costing tests? That just seems to add a lot of unnecessary overhead to me.

bknueven replied:
I'm happy with the tests as they are because the costing tests are *not* currently centralized in `watertap/costing/tests/`.

Looking at the code I'm not sure centralizing *is* the right idea -- perhaps it should be with the unit test as these are.

bknueven commented:
> @OOAmusat Ben's response was: It’s these lines: https://github.com/watertap-org/watertap/pull/515/files#diff-b8409e0826e208143b2dc914e86c9f79da1e396679bd80ab49edb430ead318e1R201-R236
> 
> These new parameters appear on every costing block. Therefore we should expect an increase in the number of Vars for any flowsheet using the watertap costing package.

That said, none of the other flowsheets *check* the number of variables. I would be okay with eliminating the check altogether. And the LSRRO tests will need to be re-done in the near future anyways.

OOAmusat committed:
Converting each cost function to separate method

OOAmusat committed:
Fix LSRRO tests

Remove ``number_of_variables`` and ``number_of_constraints`` checks from tests.

OOAmusat committed:
Merge branch 'main' into crystallization

bknueven commented on watertap/costing/watertap_costing_package.py:
```diff
@@ -586,6 +629,84 @@ def cost_caoh2_mixer(blk):
+        elif cost_type == CrystallizerCostType.volume_basis:
+            WaterTAPCostingData.cost_crystallizer_by_volume(blk)
+        else:
+            raise ConfigurationError(
+                f"{blk.unit_model.name} received invalid argument for cost_type:"
+                f" {cost_type}. Argument must be a member of the CrystallizerCostType Enum."
+            )
```
I do test that this error is raised for the other methods here: https://github.com/watertap-org/watertap/blob/main/watertap/costing/tests/test_watertap_costing_package.py

OOAmusat replied:
Test added.

bknueven commented on watertap/unit_models/crystallizer.py:
```diff
@@ -664,6 +664,10 @@ def initialize(
 
+    def get_costing(self, module=None, **kwargs):
+        self.costing = Block()
+        module.Crystallization_costing(self.costing)
```
I'm not sure this code is strictly necessary -- we do not currently use the "old" IDAES costing framework anywhere in WaterTAP, and we should probably consider removing this method from our other unit models.

OOAmusat replied:
Done

bknueven reviewed (commented):
A few suggestions for increasing the coverage.

OOAmusat committed:
Test config error, remove unnecessary block call

OOAmusat committed:
Merge branch 'crystallization' of https://github.com/OOAmusat/watertap into crystallization

bknueven reviewed (approved):
I cannot comment on the two costing models, but their implementation looks good.

adam-a-a reviewed (approved):
LGTM
